# Credit Risk Resampling Techniques

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [3]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

# Read the CSV and Perform Basic Data Cleaning

In [4]:
columns = [
    "loan_amnt", "int_rate", "installment", "home_ownership",
    "annual_inc", "verification_status", "issue_d", "loan_status",
    "pymnt_plan", "dti", "delinq_2yrs", "inq_last_6mths",
    "open_acc", "pub_rec", "revol_bal", "total_acc",
    "initial_list_status", "out_prncp", "out_prncp_inv", "total_pymnt",
    "total_pymnt_inv", "total_rec_prncp", "total_rec_int", "total_rec_late_fee",
    "recoveries", "collection_recovery_fee", "last_pymnt_amnt", "next_pymnt_d",
    "collections_12_mths_ex_med", "policy_code", "application_type", "acc_now_delinq",
    "tot_coll_amt", "tot_cur_bal", "open_acc_6m", "open_act_il",
    "open_il_12m", "open_il_24m", "mths_since_rcnt_il", "total_bal_il",
    "il_util", "open_rv_12m", "open_rv_24m", "max_bal_bc",
    "all_util", "total_rev_hi_lim", "inq_fi", "total_cu_tl",
    "inq_last_12m", "acc_open_past_24mths", "avg_cur_bal", "bc_open_to_buy",
    "bc_util", "chargeoff_within_12_mths", "delinq_amnt", "mo_sin_old_il_acct",
    "mo_sin_old_rev_tl_op", "mo_sin_rcnt_rev_tl_op", "mo_sin_rcnt_tl", "mort_acc",
    "mths_since_recent_bc", "mths_since_recent_inq", "num_accts_ever_120_pd", "num_actv_bc_tl",
    "num_actv_rev_tl", "num_bc_sats", "num_bc_tl", "num_il_tl",
    "num_op_rev_tl", "num_rev_accts", "num_rev_tl_bal_gt_0",
    "num_sats", "num_tl_120dpd_2m", "num_tl_30dpd", "num_tl_90g_dpd_24m",
    "num_tl_op_past_12m", "pct_tl_nvr_dlq", "percent_bc_gt_75", "pub_rec_bankruptcies",
    "tax_liens", "tot_hi_cred_lim", "total_bal_ex_mort", "total_bc_limit",
    "total_il_high_credit_limit", "hardship_flag", "debt_settlement_flag"
]

target = ["loan_status"]

In [5]:
# Load the data
file_path = Path('Resources/LoanStats_2019Q1.csv')
df = pd.read_csv(file_path, skiprows=1)[:-2]
df = df.loc[:, columns].copy()

# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')

# Drop the null rows
df = df.dropna()

# Remove the `Issued` loan status
issued_mask = df['loan_status'] != 'Issued'
df = df.loc[issued_mask]

# convert interest rate to numerical
df['int_rate'] = df['int_rate'].str.replace('%', '')
df['int_rate'] = df['int_rate'].astype('float') / 100


# Convert the target column values to low_risk and high_risk based on their values
x = {'Current': 'low_risk'}   
df = df.replace(x)

x = dict.fromkeys(['Late (31-120 days)', 'Late (16-30 days)', 'Default', 'In Grace Period'], 'high_risk')    
df = df.replace(x)

df.reset_index(inplace=True, drop=True)

df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

In [6]:
df.columns

Index(['loan_amnt', 'int_rate', 'installment', 'home_ownership', 'annual_inc',
       'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv',
       'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'next_pymnt_d', 'collections_12_mths_ex_med',
       'policy_code', 'application_type', 'acc_now_delinq', 'tot_coll_amt',
       'tot_cur_bal', 'open_acc_6m', 'open_act_il', 'open_il_12m',
       'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il', 'il_util',
       'open_rv_12m', 'open_rv_24m', 'max_bal_bc', 'all_util',
       'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_si

In [7]:
# Binary encoding using Pandas (multiple columns)
loans_binary_encoded = pd.get_dummies(df, columns=["home_ownership","verification_status","pymnt_plan","issue_d","initial_list_status","next_pymnt_d","application_type","hardship_flag","debt_settlement_flag"])
loans_binary_encoded.head()

loan_amnt  int_rate  installment  annual_inc loan_status    dti  \
0    10500.0    0.1719       375.35     66000.0    low_risk  27.24   
1    25000.0    0.2000       929.09    105000.0    low_risk  20.23   
2    20000.0    0.2000       529.88     56000.0    low_risk  24.26   
3    10000.0    0.1640       353.55     92000.0    low_risk  31.44   
4    22000.0    0.1474       520.39     52000.0    low_risk  18.76   

   delinq_2yrs  inq_last_6mths  open_acc  pub_rec  ...  issue_d_Jan-2019  \
0          0.0             0.0       8.0      0.0  ...                 0   
1          0.0             0.0      17.0      1.0  ...                 0   
2          0.0             0.0       8.0      0.0  ...                 0   
3          0.0             1.0      10.0      1.0  ...                 0   
4          0.0             1.0      14.0      0.0  ...                 0   

   issue_d_Mar-2019  initial_list_status_f  initial_list_status_w  \
0                 1                      0                      1   
1                 1                      0                      1   
2                 1                      0                      1   
3                 1                      0                      1   
4                 1                      0                      1   

   next_pymnt_d_Apr-2019  next_pymnt_d_May-2019  application_type_Individual  \
0                      0                      1                            1   
1                      0                      1                            1   
2                      0                      1                            1   
3                      0                      1                            1   
4                      0                      1                            1   

   application_type_Joint App  hardship_flag_N  debt_settlement_flag_N  
0                           0                1                       1  
1                           0                1                       1  
2                           0                1                       1  
3                           0                1                       1  
4                           0                1                       1  

[5 rows x 96 columns]

In [8]:
loans_binary_encoded.columns

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'loan_status',
       'dti', 'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec',
       'revol_bal', 'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_

# Split the Data into Training and Testing

In [9]:
# Create our features
X = loans_binary_encoded.drop(columns=["loan_status"])

In [10]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  issue_d_Jan-2019  issue_d_Mar-2019  \
count   68817.000000  ...      68817.000000      68817.000000   
mean    17604.142828  ...          0.451066          0.177238   
std     21835.880400  ...          0.497603          0.381873   
min         0.000000  ...          0.000000          0.000000   
25%      6293.000000  ...          0.000000          0.000000   
50%     12068.000000  ...          0.000000          0.000000   
75%     21735.000000  ...          1.000000          0.000000   
max    587191.000000  ...          1.000000          1.000000   

       initial_list_status_f  initial_list_status_w  next_pymnt_d_Apr-2019  \
count           68817.000000           68817.000000           68817.000000   
mean                0.123879               0.876121               0.383161   
std                 0.329446               0.329446               0.486161   
min                 0.000000               0.000000               0.000000   
25%                 0.000000               1.000000               0.000000   
50%                 0.000000               1.000000               0.000000   
75%                 0.000000               1.000000               1.000000   
max                 1.000000               1.000000               1.000000   

       next_pymnt_d_May-2019  application_type_Individual  \
count           68817.000000                 68817.000000   
mean                0.616839                     0.860340   
std                 0.486161                     0.346637   
min                 0.000000                     0.000000   
25%                 0.000000                     1.000000   
50%                 1.000000                     1.000000   
75%                 1.000000                     1.000000   
max                 1.000000                     1.000000   

       application_type_Joint App  hardship_flag_N  debt_settlement_flag_N  
count                68817.000000          68817.0                 68817.0  
mean                     0.139660              1.0                     1.0  
std                      0.346637              0.0                     0.0  
min                      0.000000              1.0                     1.0  
25%                      0.000000              1.0                     1.0  
50%                      0.000000              1.0                     1.0  
75%                      0.000000              1.0                     1.0  
max                      1.000000              1.0                     1.0  

[8 rows x 95 columns]

In [11]:
# Create our target
y = pd.DataFrame(df["loan_status"])
y.head()

loan_status
0    low_risk
1    low_risk
2    low_risk
3    low_risk
4    low_risk

In [12]:
y.groupby(["loan_status"])["loan_status"].count()

loan_status
high_risk      347
low_risk     68470
Name: loan_status, dtype: int64

In [13]:
# Check the balance of our target values
y['loan_status'].value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [14]:
# Create X_train, X_test, y_train, y_test
# Normal train-test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y['loan_status'], random_state=1)
Counter(y_train)

Counter({'low_risk': 51366, 'high_risk': 246})

# Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble AdaBoost classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Print the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier onely, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

# Balanced Random Forest Classifier

In [15]:
# Train the model 
from imblearn.ensemble import BalancedRandomForestClassifier
brf = BalancedRandomForestClassifier(n_estimators=100, random_state=1)
brf.fit(X_train, y_train) 

BalancedRandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                               criterion='gini', max_depth=None,
                               max_features='auto', max_leaf_nodes=None,
                               max_samples=None, min_impurity_decrease=0.0,
                               min_samples_leaf=2, min_samples_split=2,
                               min_weight_fraction_leaf=0.0, n_estimators=100,
                               n_jobs=None, oob_score=False, random_state=1,
                               replacement=False, sampling_strategy='auto',
                               verbose=0, warm_start=False)

In [16]:
# Generate predictions
y_pred = brf.predict(X_test)

# Calculate the balanced accuracy score
balanced_accuracy_score(y_test, y_pred)

0.7699825875945874

In [17]:
# Display the confusion matrix
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[   68,    33],
       [ 2280, 14824]], dtype=int64)

In [18]:
# Print the imbalanced classification report
from imblearn.metrics import classification_report_imbalanced

print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.03      0.67      0.87      0.06      0.76      0.57       101
   low_risk       1.00      0.87      0.67      0.93      0.76      0.59     17104

avg / total       0.99      0.87      0.67      0.92      0.76      0.59     17205



In [19]:
# Calculate feature importance in the Balanced Random Forest model.
# The output from this code returns an array of scores for the features in the X_test set, whose sum equals 1.0
brf.feature_importances_

array([0.01095956, 0.03808461, 0.01819198, 0.01358939, 0.01484838,
       0.00235351, 0.00435994, 0.00767562, 0.00098932, 0.01476718,
       0.01002912, 0.01668061, 0.01491561, 0.05595581, 0.07360596,
       0.08054915, 0.06003119, 0.00542627, 0.        , 0.        ,
       0.05345048, 0.        , 0.        , 0.        , 0.00347345,
       0.01285407, 0.00993662, 0.00529261, 0.00392328, 0.00695064,
       0.01541558, 0.01295109, 0.01151492, 0.00316092, 0.00718923,
       0.01292445, 0.01188838, 0.01297629, 0.00528409, 0.00544154,
       0.01133732, 0.0065587 , 0.01574347, 0.01565681, 0.01155517,
       0.        , 0.        , 0.01163733, 0.01284246, 0.00802334,
       0.00968104, 0.00621848, 0.01207183, 0.01862107, 0.00221038,
       0.00810471, 0.00752672, 0.00634078, 0.00818251, 0.0062795 ,
       0.00900861, 0.01145948, 0.00791808, 0.00947732, 0.        ,
       0.        , 0.00102456, 0.00753443, 0.00811721, 0.00553697,
       0.00106443, 0.        , 0.01464703, 0.01225645, 0.01408

In [20]:
# We can sort the features by their importance.
sorted(zip(brf.feature_importances_, X.columns), reverse=True)

[(0.08054915386941022, 'total_rec_prncp'),
 (0.07360595859725914, 'total_pymnt_inv'),
 (0.06003118785664784, 'total_rec_int'),
 (0.055955813965163706, 'total_pymnt'),
 (0.05345048388222187, 'last_pymnt_amnt'),
 (0.03808461133728464, 'int_rate'),
 (0.02111249719688521, 'issue_d_Jan-2019'),
 (0.018621066213503325, 'mths_since_recent_inq'),
 (0.018191977501513017, 'installment'),
 (0.01668060759134326, 'out_prncp'),
 (0.015743471017779034, 'avg_cur_bal'),
 (0.01565680699266943, 'bc_open_to_buy'),
 (0.01541558303173369, 'mths_since_rcnt_il'),
 (0.014915607895463924, 'out_prncp_inv'),
 (0.014848377322145442, 'dti'),
 (0.014767179686989908, 'revol_bal'),
 (0.014647032726193183, 'tot_hi_cred_lim'),
 (0.014565229805131573, 'issue_d_Mar-2019'),
 (0.014080044843290288, 'total_bc_limit'),
 (0.013589387370574305, 'annual_inc'),
 (0.012976286600397366, 'total_rev_hi_lim'),
 (0.012951090627881958, 'total_bal_il'),
 (0.012924454776333234, 'max_bal_bc'),
 (0.012854069067531457, 'tot_cur_bal'),
 (0.012

## Analysis:

The accuracy score is 0.77 which means that the model is correct 77% of the time. From the imbalanced classification report, the precision is 1 for low-risk(majority class) and 0.03 for high-risk(minority class) loans which is not a good value to rely upon. The recall for high-risk loans is less than the low-risk loans but F1 score is still very poor for the high-risk loans. Hence, this model is also not a good predictor to assess the credit risk. This model can rather be used to access low-risk loans better than the high-risk loans.

# Easy Ensemble Classifier

In [21]:
# Train the model 
from imblearn.ensemble import EasyEnsembleClassifier
eec = EasyEnsembleClassifier(n_estimators=100, random_state=1)
eec.fit(X_train, y_train)

EasyEnsembleClassifier(base_estimator=None, n_estimators=100, n_jobs=None,
                       random_state=1, replacement=False,
                       sampling_strategy='auto', verbose=0, warm_start=False)

In [22]:
# Generate predictions
y_pred = eec.predict(X_test)

# Calculate the balanced accuracy score
balanced_accuracy_score(y_test, y_pred)

0.9328586214561587

In [23]:
# Display the confusion matrix
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[   93,     8],
       [  942, 16162]], dtype=int64)

In [24]:
# Print the imbalanced classification report
from imblearn.metrics import classification_report_imbalanced

print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.09      0.92      0.94      0.16      0.93      0.87       101
   low_risk       1.00      0.94      0.92      0.97      0.93      0.87     17104

avg / total       0.99      0.94      0.92      0.97      0.93      0.87     17205



## Analysis:

The accuracy score is 0.93 which means that the model is correct 93% of the time. From the imbalanced classification report, the precision is 1 for low-risk(majority class) and 0.09 for high-risk(minority class) loans which is not a good value to rely upon. The recall for both high-risk loans and the low-risk loans is very high. Although F1 score is 0.16, but it is better than all the above models for the high-risk loans. Hence, this model may be used to assess the credit risk bit it is still not a good predictor. 